File: sdk/runanywhere-flutter/packages/runanywhere/lib/features/cloud/openai_compatible_provider.dart
/// OpenAI Compatible Provider
///
/// Built-in cloud provider for OpenAI-compatible APIs.
/// Works with OpenAI, Groq, Together, Ollama, vLLM, etc.
///
/// Mirrors Swift OpenAICompatibleProvider from Features/Cloud/OpenAICompatibleProvider.swift
library openai_compatible_provider;

import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'package:runanywhere/features/cloud/cloud_provider.dart';
import 'package:runanywhere/features/cloud/cloud_types.dart';
import 'package:runanywhere/foundation/logging/sdk_logger.dart';

// MARK: - OpenAI Compatible Provider

/// Cloud provider for any OpenAI-compatible chat completions API.
///
/// Supports both streaming (SSE) and non-streaming responses.
///
/// ```dart
/// // OpenAI
/// final openai = OpenAICompatibleProvider(apiKey: 'sk-...', model: 'gpt-4o-mini');
///
/// // Groq
/// final groq = OpenAICompatibleProvider(
///   apiKey: 'gsk_...',
///   model: 'llama-3.1-8b-instant',
///   baseURL: Uri.parse('https://api.groq.com/openai/v1'),
/// );
///
/// // Local Ollama
/// final ollama = OpenAICompatibleProvider(
///   model: 'llama3.2',
///   baseURL: Uri.parse('http://localhost:11434/v1'),
/// );
/// ```
class OpenAICompatibleProvider implements CloudProvider {
  // MARK: - CloudProvider

  @override
  final String providerId;

  @override
  final String displayName;

  // MARK: - Configuration

  final String? _apiKey;
  final String _model;
  final Uri _baseURL;
  final Map<String, String> _additionalHeaders;
  final HttpClient _httpClient;
  final SDKLogger _logger;

  // MARK: - Init

  /// Create an OpenAI-compatible provider.
  ///
  /// - [providerId] Unique ID (default: auto-generated from base URL)
  /// - [displayName] Human-readable name
  /// - [apiKey] API key (null for local providers like Ollama)
  /// - [model] Default model to use
  /// - [baseURL] API base URL (default: OpenAI)
  /// - [additionalHeaders] Extra headers to send with every request
  OpenAICompatibleProvider({
    String? providerId,
    String? displayName,
    String? apiKey,
    required String model,
    Uri? baseURL,
    Map<String, String> additionalHeaders = const {},
  })  : _apiKey = apiKey,
        _model = model,
        _baseURL = baseURL ?? Uri.parse('https://api.openai.com/v1'),
        _additionalHeaders = additionalHeaders,
        _httpClient = HttpClient(),
        _logger = SDKLogger('OpenAICompatibleProvider'),
        providerId = providerId ??
            'openai-compat-${(baseURL ?? Uri.parse("https://api.openai.com/v1")).host}',
        displayName = displayName ??
            'OpenAI Compatible (${(baseURL ?? Uri.parse("https://api.openai.com/v1")).host})';

  // MARK: - CloudProvider Implementation

  @override
  Future<CloudGenerationResult> generate(
    String prompt,
    CloudGenerationOptions options,
  ) async {
    final stopwatch = Stopwatch()..start();

    final messages = _buildMessages(prompt, options);
    final requestBody = _buildRequestBody(messages, options, stream: false);

    final data = await _performRequest(requestBody);
    stopwatch.stop();

    final response = jsonDecode(data) as Map<String, dynamic>;
    final choices = response['choices'] as List<dynamic>? ?? [];
    final text = choices.isNotEmpty
        ? ((choices[0] as Map<String, dynamic>)['message']
                as Map<String, dynamic>)['content'] as String? ??
            ''
        : '';

    final usage = response['usage'] as Map<String, dynamic>?;
    final promptTokens = usage?['prompt_tokens'] as int? ?? 0;
    final completionTokens = usage?['completion_tokens'] as int? ?? 0;

    return CloudGenerationResult(
      text: text,
      inputTokens: promptTokens,
      outputTokens: completionTokens,
      latencyMs: stopwatch.elapsedMilliseconds.toDouble(),
      providerId: providerId,
      model: options.model,
    );
  }

  @override
  Stream<String> generateStream(
    String prompt,
    CloudGenerationOptions options,
  ) {
    final controller = StreamController<String>();

    _streamGeneration(prompt, options, controller);

    return controller.stream;
  }

  @override
  Future<bool> isAvailable() async {
    try {
      final url = _baseURL.resolve('models');
      final request = await _httpClient.getUrl(url);
      request.headers.set('Content-Type', 'application/json');
      if (_apiKey != null) {
        request.headers.set('Authorization', 'Bearer $_apiKey');
      }

      final response =
          await request.close().timeout(const Duration(seconds: 5));
      return response.statusCode == 200;
    } catch (e) {
      _logger.debug('Provider availability check failed: $e');
      return false;
    }
  }

  // MARK: - Internal Helpers

  Future<void> _streamGeneration(
    String prompt,
    CloudGenerationOptions options,
    StreamController<String> controller,
  ) async {
    try {
      final messages = _buildMessages(prompt, options);
      final requestBody = _buildRequestBody(messages, options, stream: true);
      final request = await _buildHttpRequest(requestBody);

      final response = await request.close();

      if (response.statusCode < 200 || response.statusCode >= 300) {
        controller.addError(
          CloudProviderException.httpError(response.statusCode),
        );
        await controller.close();
        return;
      }

      // Parse SSE stream
      await response
          .transform(utf8.decoder)
          .transform(const LineSplitter())
          .forEach((line) {
        if (line.startsWith('data: ')) {
          final data = line.substring(6);
          if (data == '[DONE]') return;

          try {
            final chunk = jsonDecode(data) as Map<String, dynamic>;
            final choices = chunk['choices'] as List<dynamic>? ?? [];
            if (choices.isNotEmpty) {
              final delta = (choices[0] as Map<String, dynamic>)['delta']
                  as Map<String, dynamic>?;
              final content = delta?['content'] as String?;
              if (content != null) {
                controller.add(content);
              }
            }
          } catch (e) {
            // Skip malformed chunks
            _logger.debug('Skipping malformed SSE chunk: $e');
          }
        }
      });

      await controller.close();
    } catch (e) {
      controller.addError(e);
      await controller.close();
    }
  }

  List<Map<String, String>> _buildMessages(
    String prompt,
    CloudGenerationOptions options,
  ) {
    if (options.messages != null && options.messages!.isNotEmpty) {
      return options.messages!.map((m) => m.toJson()).toList();
    }

    final msgs = <Map<String, String>>[];
    if (options.systemPrompt != null) {
      msgs.add({'role': 'system', 'content': options.systemPrompt!});
    }
    msgs.add({'role': 'user', 'content': prompt});
    return msgs;
  }

  Map<String, dynamic> _buildRequestBody(
    List<Map<String, String>> messages,
    CloudGenerationOptions options, {
    required bool stream,
  }) {
    return {
      'model': options.model,
      'messages': messages,
      'max_tokens': options.maxTokens,
      'temperature': options.temperature,
      'stream': stream,
    };
  }

  Future<HttpClientRequest> _buildHttpRequest(
    Map<String, dynamic> body,
  ) async {
    final url = _baseURL.resolve('chat/completions');
    final request = await _httpClient.postUrl(url);

    request.headers.set('Content-Type', 'application/json');

    if (_apiKey != null) {
      request.headers.set('Authorization', 'Bearer $_apiKey');
    }

    for (final entry in _additionalHeaders.entries) {
      request.headers.set(entry.key, entry.value);
    }

    request.write(jsonEncode(body));
    return request;
  }

  Future<String> _performRequest(Map<String, dynamic> body) async {
    final request = await _buildHttpRequest(body);
    final response = await request.close();

    if (response.statusCode < 200 || response.statusCode >= 300) {
      throw CloudProviderException.httpError(response.statusCode);
    }

    return await response.transform(utf8.decoder).join();
  }
}

File: sdk/runanywhere-swift/Sources/RunAnywhere/Features/Cloud/OpenAICompatibleProvider.swift
//
//  OpenAICompatibleProvider.swift
//  RunAnywhere SDK
//
//  Built-in cloud provider for OpenAI-compatible APIs.
//  Works with OpenAI, Groq, Together, Ollama, vLLM, etc.
//

import Foundation

// MARK: - OpenAI Compatible Provider

/// Cloud provider for any OpenAI-compatible chat completions API.
///
/// Supports both streaming (SSE) and non-streaming responses.
///
/// ```swift
/// // OpenAI
/// let openai = OpenAICompatibleProvider(apiKey: "sk-...", model: "gpt-4o-mini")
///
/// // Groq
/// let groq = OpenAICompatibleProvider(
///     apiKey: "gsk_...",
///     model: "llama-3.1-8b-instant",
///     baseURL: URL(string: "https://api.groq.com/openai/v1")!
/// )
///
/// // Local Ollama
/// let ollama = OpenAICompatibleProvider(
///     model: "llama3.2",
///     baseURL: URL(string: "http://localhost:11434/v1")!
/// )
/// ```
public final class OpenAICompatibleProvider: CloudProvider, @unchecked Sendable {

    // MARK: - CloudProvider

    public let providerId: String
    public let displayName: String

    // MARK: - Configuration

    private let apiKey: String?
    private let model: String
    private let baseURL: URL
    private let additionalHeaders: [String: String]
    private let session: URLSession

    // MARK: - Init

    /// Create an OpenAI-compatible provider.
    ///
    /// - Parameters:
    ///   - providerId: Unique ID (default: auto-generated from base URL)
    ///   - displayName: Human-readable name
    ///   - apiKey: API key (nil for local providers like Ollama)
    ///   - model: Default model to use
    ///   - baseURL: API base URL (default: OpenAI)
    ///   - additionalHeaders: Extra headers to send with every request
    public init(
        providerId: String? = nil,
        displayName: String? = nil,
        apiKey: String? = nil,
        model: String,
        baseURL: URL = URL(string: "https://api.openai.com/v1")!,
        additionalHeaders: [String: String] = [:]
    ) {
        self.apiKey = apiKey
        self.model = model
        self.baseURL = baseURL
        self.additionalHeaders = additionalHeaders
        self.providerId = providerId ?? "openai-compat-\(baseURL.host ?? "local")"
        self.displayName = displayName ?? "OpenAI Compatible (\(baseURL.host ?? "local"))"
        self.session = URLSession(configuration: .default)
    }

    // MARK: - CloudProvider Implementation

    public func generate(
        prompt: String,
        options: CloudGenerationOptions
    ) async throws -> CloudGenerationResult {
        let startTime = Date()

        let messages = buildMessages(prompt: prompt, options: options)
        let requestBody = buildRequestBody(messages: messages, options: options, stream: false)

        let data = try await performRequest(body: requestBody)

        let response = try JSONDecoder().decode(ChatCompletionResponse.self, from: data)

        let latencyMs = Date().timeIntervalSince(startTime) * 1000
        let text = response.choices.first?.message.content ?? ""

        return CloudGenerationResult(
            text: text,
            inputTokens: response.usage?.promptTokens ?? 0,
            outputTokens: response.usage?.completionTokens ?? 0,
            latencyMs: latencyMs,
            providerId: providerId,
            model: options.model,
            estimatedCostUSD: nil
        )
    }

    public func generateStream(
        prompt: String,
        options: CloudGenerationOptions
    ) -> AsyncThrowingStream<String, Error> {
        AsyncThrowingStream { continuation in
            Task {
                do {
                    let messages = buildMessages(prompt: prompt, options: options)
                    let requestBody = buildRequestBody(messages: messages, options: options, stream: true)
                    let request = try buildURLRequest(body: requestBody)

                    let (bytes, response) = try await session.bytes(for: request)

                    guard let httpResponse = response as? HTTPURLResponse,
                          (200...299).contains(httpResponse.statusCode) else {
                        let statusCode = (response as? HTTPURLResponse)?.statusCode ?? -1
                        throw CloudProviderError.httpError(statusCode: statusCode)
                    }

                    // Parse SSE stream
                    for try await line in bytes.lines {
                        if line.hasPrefix("data: ") {
                            let data = String(line.dropFirst(6))
                            if data == "[DONE]" { break }

                            if let jsonData = data.data(using: .utf8),
                               let chunk = try? JSONDecoder().decode(ChatCompletionChunk.self, from: jsonData),
                               let content = chunk.choices.first?.delta.content {
                                continuation.yield(content)
                            }
                        }
                    }

                    continuation.finish()
                } catch {
                    continuation.finish(throwing: error)
                }
            }
        }
    }

    public func isAvailable() async -> Bool {
        // Simple health check: try to list models or just check connectivity
        guard let url = URL(string: "\(baseURL)/models") else { return false }
        var request = URLRequest(url: url)
        request.timeoutInterval = 5
        if let apiKey = apiKey {
            request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        }
        do {
            let (_, response) = try await session.data(for: request)
            return (response as? HTTPURLResponse)?.statusCode == 200
        } catch {
            return false
        }
    }

    // MARK: - Internal Helpers

    private func buildMessages(prompt: String, options: CloudGenerationOptions) -> [[String: String]] {
        if let messages = options.messages {
            return messages.map { ["role": $0.role, "content": $0.content] }
        }

        var msgs: [[String: String]] = []
        if let systemPrompt = options.systemPrompt {
            msgs.append(["role": "system", "content": systemPrompt])
        }
        msgs.append(["role": "user", "content": prompt])
        return msgs
    }

    private func buildRequestBody(
        messages: [[String: String]],
        options: CloudGenerationOptions,
        stream: Bool
    ) -> [String: Any] {
        var body: [String: Any] = [
            "model": options.model,
            "messages": messages,
            "max_tokens": options.maxTokens,
            "temperature": options.temperature,
            "stream": stream,
        ]
        return body
    }

    private func buildURLRequest(body: [String: Any]) throws -> URLRequest {
        guard let url = URL(string: "\(baseURL)/chat/completions") else {
            throw CloudProviderError.invalidURL
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        if let apiKey = apiKey {
            request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        }

        for (key, value) in additionalHeaders {
            request.setValue(value, forHTTPHeaderField: key)
        }

        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        return request
    }

    private func performRequest(body: [String: Any]) async throws -> Data {
        let request = try buildURLRequest(body: body)
        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse,
              (200...299).contains(httpResponse.statusCode) else {
            let statusCode = (response as? HTTPURLResponse)?.statusCode ?? -1
            throw CloudProviderError.httpError(statusCode: statusCode)
        }

        return data
    }
}

// MARK: - Cloud Provider Errors

/// Errors from cloud provider operations
public enum CloudProviderError: Error, LocalizedError, Sendable {
    case invalidURL
    case httpError(statusCode: Int)
    case noProviderRegistered
    case providerNotFound(id: String)
    case providerUnavailable(id: String)
    case decodingError(String)
    case budgetExceeded(currentUSD: Double, capUSD: Double)
    case latencyTimeout(maxMs: UInt32, actualMs: Double)

    public var errorDescription: String? {
        switch self {
        case .invalidURL: return "Invalid cloud provider URL"
        case .httpError(let code): return "Cloud API returned HTTP \(code)"
        case .noProviderRegistered: return "No cloud provider registered"
        case .providerNotFound(let id): return "Cloud provider not found: \(id)"
        case .providerUnavailable(let id): return "Cloud provider unavailable: \(id)"
        case .decodingError(let msg): return "Failed to decode cloud response: \(msg)"
        case .budgetExceeded(let current, let cap):
            return "Cloud budget exceeded: $\(String(format: "%.4f", current)) / $\(String(format: "%.4f", cap)) cap"
        case .latencyTimeout(let maxMs, let actualMs):
            return "On-device latency timeout: \(String(format: "%.0f", actualMs))ms exceeded \(maxMs)ms limit"
        }
    }
}

// MARK: - OpenAI API Response Types

private struct ChatCompletionResponse: Decodable {
    let choices: [Choice]
    let usage: Usage?

    struct Choice: Decodable {
        let message: Message
    }

    struct Message: Decodable {
        let content: String?
    }

    struct Usage: Decodable {
        let promptTokens: Int
        let completionTokens: Int

        enum CodingKeys: String, CodingKey {
            case promptTokens = "prompt_tokens"
            case completionTokens = "completion_tokens"
        }
    }
}

private struct ChatCompletionChunk: Decodable {
    let choices: [Choice]

    struct Choice: Decodable {
        let delta: Delta
    }

    struct Delta: Decodable {
        let content: String?
    }
}
